app/models/user.rb

```diff
@@ -0,0 +1,27 @@
+class User < ActiveRecord::Base
+  include Slug
+  sluggable_column :lastname
+  before_save :generate_slug
+
+  validates_presence_of :firstname
+  validates_presence_of :lastname
+
+  validates :email, presence: true, uniqueness: true
+  validates :password, presence: true, length: {minimum: 6}, on: :create
+
+
+  has_secure_password
+
+
+  before_save do 
+    generate_token(:auth_token)
+  end
+
+  # generate random token for login
+  def generate_token(column)
+    begin
+      self[column] = SecureRandom.urlsafe_base64
+    end while User.exists?(column == self[column])
```

Comment:
This is not doing what you think it is. You are doing this:

```ruby
User.exists?(column == self[column])
```

Which is equivalent to:

```ruby
User.exists?(:auth_token == 'some_token')
```

which is equivalent to:

```ruby
User.exists?(false)
```

This is what you want to do:

```ruby
User.exists?(column => self[column])
```

You need to provide a hash which has the column name and value you are looking for.

http://apidock.com/rails/ActiveRecord/Base/exists%3F/class